internal/p2p/rqueue.go

```diff
@@ -0,0 +1,111 @@
+package p2p
+
+import (
+	"container/heap"
+	"context"
+	"sort"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+type simpleQueue struct {
+	input   chan Envelope
+	output  chan Envelope
+	closeFn func()
+	closeCh <-chan struct{}
+
+	maxSize int
+	chDescs []*ChannelDescriptor
+}
+
+func newSimplePriorityQueue(ctx context.Context, size int, chDescs []*ChannelDescriptor) *simpleQueue {
+	if size%2 != 0 {
+		size++
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	q := &simpleQueue{
+		input:   make(chan Envelope, size*2),
+		output:  make(chan Envelope, size/2),
+		maxSize: size * size,
```

Comment:
There's some subtle size arithmetic going on here and below, and I'm not sure I understand the motivation for it. If the idea is to have a heap with a certain capacity, shouldn't that capacity be the (actual) maxSize?

Reply:
magic numbers are magic. I think we've generally been incredibly (and needlessly) conservative with these kinds of numbers, and given that our previous queuing approach was unbounded, I think squaring all the specified queue sizes is definitely reasonable

Reply:
Ah yes, I have no objection to using a bigger value, I'm just not understanding why we need to. Is this just to avoid having to resize the buffer as often?

Reply:
Right—agreed. But if the idea is just to have a larger buffer, shouldn't we do that by setting the maxSize bigger rather than squaring? (If the objective is, as it seems, to avoid having to reallocate arrays as often, that seems fine, I just couldn't tell)

Reply:
I'm not particularly worried about array resizing. 

I think it's also the case that I wanted this change to impact the queue sizes (which are defined on the channel descriptors (I think?)